src/hooks/useDownload.ts

```diff
@@ -82,26 +85,32 @@ async function internal<T>({ src, headers, logger, dst }: DownloadOptions): Prom
       let n = 0
       return [dst, reader.pipeThrough(new TransformStream({
         transform: (buf, controller) => {
-          let s = txt
-          if (!sz) {
-            s += ` ${pretty_size(n)}`
+          n += buf.length
+          if (json) {
+            console.log({status: "downloading", "received": n, "content-size": sz })
+          } else if (!sz) {
+            (logger as Logger).replace(`${txt} ${pretty_size(n)}`)
           } else {
-            n += buf.length
+            let s = txt
             if (n < sz) {
               let pc = n / sz * 100;
               pc = pc < 1 ? Math.round(pc) : Math.floor(pc);  // don’t say 100% at 99.5%
               s += ` ${pc}%`
             } else {
               s = teal('extracting')
             }
+            (logger as Logger).replace(s)
           }
-          (logger as Logger).replace(s)
           controller.enqueue(buf)
       }}))]
     }
   }
   case 304:
-    logger.replace(`cache: ${teal('hit')}`)
+    if (json) {
```

Comment:
I wonder if there's a good way to create a smarter logger to avoid all these if statements?  So you could instantiate a `Logger` or a `JsonLogger` and it would know how to format stuff.  Although to make that work you'd want to send events to the logger instead of raw text...

Not something for now, but maybe in the future if we get a lot of if/else's

Reply:
Yeah indeed. There's certainly a way to make the code cleaner. Both files could just have a `SpecificLogger` class with functions for each step that take the parameters they need to log for the logging type they are doing. Possibly something even more generic but I think it would end up being abstration masturbation.

I considered it but didn't want to spend longer on it and went for the quick win.